keps/sig-network/4963-kube-proxy-flowtables-fastpath/README.md

```diff
@@ -0,0 +1,463 @@
+any other network applications that depend on the packets going through the network stack (monitoring per example) we'll not be able to see the connection data. The feature will only
+apply the fast path on established connections, since most of the network applications are statefuls, usually is safe to think that once a connection is established no additional operations are required on it.
+
+## Design Details
+
+This feature will only work with kube-proxy nftables mode.
+
+Users will be able to opt-in to Service traffic acceleration by passing a CEL expression using the flag `--accelerated-interface-expression` or the configuration option `AcceleratedInterfaceExpression` to match the network interfaces in the node that are subjet to Service traffic acceleration. The absence of a CEL expression disables the feature.
+
+Kube-proxy will create a `flowtable` in the kube-proxy table with the name `kube-proxy-flowtable` and will monitor the network interfaces in the node to populate the `flowtable` with the interfaces that match the configured CEL expression.
+
+Kube-proxy will insert a rule to offload all Services established traffic in the `filter-forward` chain:
+
+```
+	tx.Add(&knftables.Rule{
+		Chain: filterForwardChain,
+		Rule: knftables.Concat(
+			"ct original", ipX, "daddr", "@", clusterIPsSet,
+			"ct state established",
+			"flow offload", "@", serviceFlowTable,
+		),
+	})
+```
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+To be added
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+Not needed
+
+##### e2e tests
+
+- Create one Service with one backend running an iperf Service
+- Run an iperf client against the Service without acceleration
+- Run an iperf client against the Service with acceleration.
+- The service with acceleration has to show a significant statistically difference on the throughput results.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers
+
+#### GA
+
+- No bug reported
+- Feedback from developers and users
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+
+### Upgrade / Downgrade Strategy
+
+kube-proxy reconcile the nftables rules so the rules will be reconciled during startup and added or removed depending on how kube-proxy is configured.
```

Comment:
FWIW when downgrading, any existing flow table entries will be preserved, so any Service connections that remain up across a kube-proxy downgrade would still be offloaded and there'd be no way to un-offload them with the older kube-proxy.

(Actually, currently that's true even if you restart the new kube-proxy with the feature disabled; you need to delete the flowtable at startup in that case.)

Reply:
hmm, I need to document and understand this better, if there is a way to cleanup we already have a `kube-proxy cleanup` command so I wonder if we can reuse it for this

Reply:
I was talking about downgrading to a release that doesn't know anything about flowtables, but I guess that's only a problem for people using it during Alpha.

Anyway, you just need to make it so that if you start nftables kube-proxy in a version that knows about flowtables, but with flowtables disabled, then it should ensure the flow table _doesn't_ exist when it does its first sync rather than ensuring that it _does_ exist.

Reply:
@danwinship I see now in the code that we flush chain by chain instead of the whole table when starting ... is there any reason for doing that? flushing the whole table allow to handle downgrade/upgrade of the table schema without further considerations

Reply:
When I first wrote the code, I thought `nft flush table ip kube-proxy` would flush the contents of sets too, which would mean that restarting kube-proxy would break any active affinity. But it turns out that it doesn't actually do that anyway; flushing a table just flushes (but does not delete) every chain within the table.

Reply:
should then we change kube-proxy to flush the entire table?

Reply:
flush still keeps the flowtable , so it is not valid

Reply:
I need to think if is possible to have an integrity check 🤔  and do a full resync